src/carousel/index.ts

```diff
@@ -0,0 +1,206 @@
+import { alias, tag, utils, Tag } from '@storefront/core';
+
+const MIN_SWIPE_DISTANCE = 20;
+
+@alias('carousel')
+@tag('gb-carousel', require('./index.html'), require('./index.css'))
+class Carousel {
+  refs: {
+    wrapper: HTMLDivElement,
+    track: HTMLDivElement
+  };
+  props: Carousel.Props = {
+    speed: 0,
+    slidestoshow: 1,
+    slidestoscroll: 1,
+    items: []
+  };
+  state: Carousel.State = {
+    currentSlide: 0,
+    transitioning: false,
+    items: []
+  };
+  animationEndCallback: NodeJS.Timer | null = null;
+
+  onMount() {
+    utils.WINDOW().addEventListener('resize', this.update);
+  }
+
+  onUpdate() {
+    this.state.items = this.cloneItems();
+  }
+
+  onUnmount() {
+    utils.WINDOW().removeEventListener('resize', this.update);
+  }
+
+  moveNext = () => this.slideHandler(this.state.currentSlide + this.props.slidestoscroll);
+
+  movePrevious = () => this.slideHandler(this.state.currentSlide - this.props.slidestoscroll);
+
+  cloneItems = () => {
+    const numCloned = this.props.slidestoshow * 2 - 1;
+    const length = this.props.items.length;
+
+    const prior = this.props.items
+      .slice(-numCloned)
+      .map((data, index) => ({ ...data, 'data-index': index - length }));
+
+    this.props.items.forEach((data, index) => (data['data-index'] = index));
+
+    const posterior = this.props.items
+      .slice(0, numCloned)
+      .map((data, index) => ({ ...data }));
+
+    return prior.concat(this.props.items).concat(posterior);
+  }
+
+  slideHandler = (slide: number) => {
+    const { speed } = this.props;
+    const from = this.state.currentSlide;
+    const length = this.props.items.length;
+    const onEdge = slide >= length || slide <= 0;
+
+    // don't need to consider on edge if speed is 0;
+    if (!speed) {
+      this.resetCurrentSlideNum(from, slide, length);
+      return;
+    }
+
+    if (!(this.state.transitioning && onEdge)) {
+      // make the transition
+      this.set({ currentSlide: slide, transitioning: true });
```

Comment:
I'd like to start the transition only after we've set up the event listeners below:
```js
if (onEdge) {
  // reset to non-cloned slide
  // tslint:disable-next-line:max-line-length
  this.animationEndCallback = setTimeout(() => this.resetToRealSlide(from, slide, length), speed);
} else {
  this.refs.track.addEventListener('transitionend', this.disableTransition, false);
}

this.set({ currentSlide: slide, transitioning: true });
```